==Supertree Tool Kit (STK)==
STK is a perl module and collection of script designed to make processing data for supertree analysis easier. It is not an automated system that blindly alters data: human intelligence is still required.

==Description==
Creating supertrees containing thousands of taxa requires hundreds of input source phylogenies. Managing and processing this data in a systematic, error-free, manner challenging and will become even more so as supertrees contain increasing numbers of taxa. Protocols for processing input source phylogenies have been proposed to ensure data quality. Therefore any tool for processing source phylogenies should enable these protocols to be followed easily.

The Supertree Tool Kit is a Perl module with associated executables scripts that aim to aid in the collection, storage and processing of input source trees for use in supertree analysis. It is therefore invaluable when creating supertrees containing thousands of taxa. We have added meta-data via XML files to each tree that contains information such as the source information for the tree (authors, journal, etc) and how the data were derived, including the analysis algorithm and character data used to carry out the original analysis.

The STK is a bioinformatics tool designed to make it easier to construct supertrees from hundreds or thousands of input source trees, whilst reducing potential errors and enabling easy sharing of such datasets.

==Files==
STK uses three types of files:

; '''NEXUS formatted tree files'''
:These are simply standard NEXUS files. Bio::Nexus does a reasonable job of reading in a lot of the variations in this format. The STK assumes these files have a <tt>.tre</tt> extension. If not, they will not be read.
; '''XML meta-data files'''
:XML is a standard for meta-data.
; '''substitute taxa files'''
:A file that contains taxa and the taxa to substitute with.

===NEXUS files===
Standard NEXUS file as read in by Bio::Nexus (Hladish et al., 2007). These can be created by a number of programs, including TreeView, Mesquite, etc. TreeView produces files that cannot be read in by Bio::Nexus, but a script is included to fix these.

===XML files===
Are of the layout:

<code>
    <SourceTree>
        <Source>
            <!-- If it's a journal, it has a journal name. Book has booktitle -->
            <Author>authors</Author>
            <Year>year</Year>
            <Title>title</Title>
            <Journal>Auk</Journal>
            <Volume></Volume>
            <Pages></Pages>
            <Booktitle/>
            <Editor/>
            <Publisher/>
        </Source>
        <Taxa fossil="none|some|all" number="n">
            <List>taxon 1</List>
            ...
            <List>taxon n</List>
        </Taxa>
        <Characters>
            <Molecular>
                <Type>type</Type>
            </Molecular>
            <Morphological/>
            <Behavioural/>
            <Other/>
        </Characters>
        <Analysis>
            <Type></Type>
        </Analysis>
        <Notes>Some notes</Notes>
        <TreeFile>A tree file</TreeFile>
    </SourceTree>
    
</code>
===Substitute taxa files===
These are of the format:

<code>
 taxon_1 = taxon_2,taxon_3
 taxon_4 = taxon_5,taxon_6,taxon_7
 taxon_8 = 
</code>
A blank after the <tt>=</tt> sign indicates a removal of a taxon. Any number of replacement taxa can be used on a single line. There must be a space either side of the <tt>=</tt> sign.

==Data Structure==
The STK was designed around a typical user who liked to keep back-up copies of data within the main dataset. As such certain directories and files are ignored when scanning a dataset. These are (note that the names are case-insensitive and italics indicate filenames, rather than directory names):

* '''original'''
* '''Adj for'''
* '''Adj_for'''
* ''temp.tre''
* ''combinedTrees''
* ''all_trees''

==Scripts==
The STK consists of a number of executable scripts. Each script can be run using: <tt>stk_script_name [options]</tt>, if the STK has been installed, otherwise they can be run using <tt>perl stk_script_namei.pl [options]</tt> from the exec directory.

The following scripts are included in the STK:

* Amalagamate Trees
* Check Data
* Check Overlap
* Check Substitutions
* Create Matrix
* Data Independence
* Data Summary
* Fix TreeFiles
* Replace Genera
* Replace Taxa
* Tree Permutation
* Search Data

More details are given in the sections below.

===Amalgamate Trees===
====Synopsis====
stk_amalgamate_trees.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing Tree/XML files. Required
   --verbose         print verbose messages
   --output          output file. Default is 'all_trees.tre' in C<dir>
   --anon            anonymise data
</code>
====Options====
; '''--dir'''
:a directory which contains tree (and XML) files. '''Required'''.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--output'''
:Where to save output. Default is 'all_trees.tre' in the <tt>dir</tt>.
; '''--anon'''
:The <tt>--anon</tt> flag to name trees tree_1...tree_n. A key file will be written in <tt>dir</tt> linking either author names or treefile names to each anonymised tree. Without this flag, tree are named after the tree authors and year published, e.g. SmithAndJones_1999 if XML files are available, or the tree filename without the suffix .tre, the prefix <tt>dir</tt>, and all non-standard characters repolaceed with underscores, e.g. trees_SmithAndJones_1999

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK.

====Description====
'''Amalgamate Trees''' creates a single tree file with all trees from <tt>dir</tt>. This is useful when sharing your data as it is the version of the data that has most compatibility with other software. If XML files are present trees will be named using the author_year convention. If XML files are not present the trees will be after the tree file name. The <tt>--anon</tt> flag to name trees tree_1...tree_n. A key file will be written in <tt>dir</tt> linking either author names or treefile names to each anonymised tree.

===Check Data===
====Synopsis====
check_data.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing XML files. Required
   --verbose         print verbose messages
   --output          direct output to file instead of STDOUT
</code>
====Options====
; '''--dir'''
:a directory which contains XML and/or tree files. '''Required'''.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--output'''
:Directs output to file instead of STDOUT. If a filename is given, output is directed there, otherwise it is directed to error.txt in the c<dir>ectory.

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK.

====Description====
'''Check Data''' checks that the data in <tt>dir</tt> is valid. For a directory of tree files this means checking that all tree files are valid NEXUS files. For a directory of XML files this means check for XML correctness and contain the minimum amount of data. For a directory of both tree and XML files the two file types are validated and cross-checked against each other. Tests carried out are:

* Matching files - each tree file should have a matching XML file (or more correctly, each XML file should have a matching tree file)
* Matching taxa - the tree and XML files should contain the same taxa

===Check Overlap===
====Synopsis====
check_overlap.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing Tree files. Required
   --verbose         print verbose messages
   --nCluster        number of taxa that trees must share. Default value is 2.
   --graphic         output a "dot" file that can be loaded into GraphViz
   --compressed      output a "dot" file where clusters are shown, rather than individual trees
</code>
====Options====
; '''--dir'''
:A directory which contains tree files.
; '''--nCluster'''
:The number of taxa required for trees to "cluster".
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--graphic'''
:Ouput a "dot" file that can be loaded into GraphViz to produce a graphical representation of the connectivity. A key file (key.csv) is also output containing the legend for the graphic
; '''--compressed'''
:Large datasets can produce "dot" files too large for neato to produce a graphic. If this is the case, using this option produce a graph of "clusters" of trees only, where a node represents a cluster of trees.

<tt>dir</tt> scans all subdirectories also, apart from those excluded by STK.

====Description====
'''Check Overlap''' checks that trees in a directory have sufficient taxonomic overlap to be used as source trees in supertree construction. Each tree must contain at least two taxon also contained in another tree (Nixon, 1996).

Check Overlap checks that a dataset fulfils this criterion. A single "cluster" of connected trees is the ideal output as this means all trees share at least two common taxa with at least one other tree. The minimum requirement for overlap to occur can be changed to create "more connected" data. Any trees that do not share the minimum number of taxa with at least one other tree should be removed fro the dataset. A warning is printed is more than one cluster is found.

If more than one cluster is found, you should use the <tt>--graphic</tt> option to visualise the data connectivity. For large datasets, the <tt>--compressed</tt> flag should also be used. The <tt>--graphic</tt>> output creates input for GraphViz, a free program that can display the overlap graphically. The <tt>--compressed</tt> flag produces output where "clusters" are represented by nodes. By definition, there will be no edges on this graph. Without this flag, each node is a tree, with edges between nodes representing minimal connectivity.

Nixon, K., Carpenter, J., 1996. On simultaneous analysis. Cladistics 12, 221-241.

===Check Substitutions===
====Synopsis====
check_substitutions.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --file            substitutions file
   --dir             directory that contains the data
   --verbose         verbose messages
   --output          output to file rather than standard out
</code>
====Options====
; '''--dir'''
:A directory of tree and optionally XML files
; '''--file'''
:A substitutions file to check
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--output'''
:File to send output to

You must specify a <tt>file</tt>. If <tt>dir</tt> is also specified the taxa within the substitution file are checked against the taxa in the dataset. Any taxa in the subs file that are not in the data set are flagged.

====Description====
'''Check Substitutions''' ensures that a substitution file as used in replace_taxa.pl is of the correct format. Additionally, if a dataset is specified, it checks the substitutions that would be made by replace_taxa.pl are alreday in the data.

For example, if your substitution file contains: taxa_1 = fred, foo, bar

taxa_1 would be replaced with a polytomy of fred, foo, bar. Specifying a dataset ensures that fred, foo, and bar are in the data - i.e. you are not introducing additional taxa into the dataset. You may '''want''' to do this and hence why replace_taxa.pl does allow this, but this is generally not a good idea, and this script checks you really mean this action.

===Clean Data===
====Synopsis====
clean_data.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing XML files. Required
   --dryrun          dry run only
   --verbose         print verbose messages
   --output          direct output to file instead of STDOUT
</code>
====Options====
; '''--dir'''
:a directory which contains XML and/or tree files. '''Required'''.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--output'''
:Directs output to file instead of STDOUT. If a filename is given, output is directed there, otherwise it is directed to error.txt in the c<dir>ectory.
; '''--dryrun'''
:Don't delete the directories, but list them instead.

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK.

====Description====
'''Clean Data''' deletes directories within the dataset that contain less than three taxa. The check is made on the XML file only. Run the check_data script first to ensure your data is as you expect. The script takes care of checking sub-directories of an author/paper directory before deleting the author/paper directory.

Note that hidden files (such as those from SVN or CVS) are a problem for the last part of this script where empty directories are deleted. If your dataset contains hidden files, then the directories will not get deleted.

===Create Directories===
====Synopsis====
create_dirs.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             path to the storage dir. Required
   --verbose         print verbose messages
   --output          output to a file, rather than standard output
   --bib             BibTex file. Required
</code>
====Options====
; '''--dir'''
:Path to the storage dir. The database will be created here. '''Required'''.
; '''--bib'''
:A BibTex file from which directory names and the base XML file will be created. '''Required'''.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages

====Description====
The script takes a BibTex file and creates the database structure. The directories will have the following format: Base_Dir/ Author_year/author_year.xml Author2_year/author_year.xml ... AuthorN_year/author_year.xml

The XML files will be created with the correct bibliographic metadata, such as journal details and paper titles.

===Create Matrix===
====Synopsis====
create_matrix.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing Tree files. Required
   --verbose         print verbose messages
   --output          output MRP file
   --format          output format. Defaults to Hening86.
</code>
====Options====
; '''--dir'''
:a directory which contains tree files. '''Required'''.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--output'''
:Output file for MRP matrix. Defaults to Matrix.tnt in <tt>dir</tt>.
; '''--format'''
:Sepcify the output format. Current Nexus and Hening86 are supported. The default is Hening86

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK.

====Description====
'''Create Matrix''' creates an MRP matrix from all tree files contained in a directory. The script uses standard MRP coding.

The script is based on that of Bininda-Emonds et al. (2005), SuperMRP.pl, but there is no support for Purvis coding, weights, etc. If you need those, use SuperMRP instead.

====References====
Bininda-Emonds, O.R.P., R.M.D. Beck, and A. Purvis. 2005. Getting to the roots of matrix representation. Systematic Biology 54(4):668-672.

===Data Independence===
====Synopsis====
data_independence.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing XML files. Required
   --verbose         print verbose messages
</code>
====Options====
; '''--dir'''
:a directory which contains XML files. '''Required'''.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK.

====Description====
'''Data Independence''' checks that source data do not actually contain the same data - that is they are independent of each other. For supertree construction, the source data must be independent, which for current protocols means that they cannot contain the same taxa (or subset of taxa) and not use the same data for the analysis. This script checks the taxonomic overlap of each study with every other study and if the taxa are the same (or one source set in wholly contained within another) checks the original source data. If these are the same, the two source studies are flagged as being potentially non-independent.

Output is a Tab-delimited file which contains the potential non-independencies. This can be loaded into Excel or similar. The data is grouped by the type of character used. The final column contains the treefile for each XML file, followed by any potential dependent trees.

This script requires XML files to function - no XML, no checks.

===Data Summary===
====Synopsis====
data_summary.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing Tree files. Required
   --taxatreematrix  create data availability matrices
   --verbose         print verbose messages
   --output          output to a file, rather than standard output
</code>
====Options====
; '''--dir'''
:A directory which contains XML/Tree files. '''Required'''.
; '''--taxatreematrix'''
:Create two file which allow data availability plots to be created.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--output'''
:Write the data summary to file, rather than standard output. Defaults to <tt>data_summary.txt</tt> in <tt>dir</tt>.

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK.

====Description====
'''Data Summary''' chreates a summary of data generated from both tree and XML files which is useful when writing manuscripts. Statistics such as the number of trees, the characters used, the taxa included, etc are output.

Note that the output depends on whether XML ''and'' Tree file are found, or just Tree files. The aim is to generate the data usually found in paper (e.g. we used 100 trees, from 185 papers, containing 500 taxa).

There is an assumption here that the XML and Treefiles live in the same directory. Therefore, specify tha topmost directory where both can be reached. This won't work where, for example, there are two directories containing treefiles and there's one directory containing XML files underneath them, e.g. data_dir/ tree_original/ tree_processed/ xml_files/

Supplying <tt>data_dir</tt> will not work correctly.

The <tt>--taxatreematrix</tt> flag creates two additional files in the <tt>data_dir</tt>: data_availability_matrix.txt taxa_tree_matrix.txt

These two files allow visualisation of the data. The former is comparable to the popular visualisation of the data availability in molecular studies where genes are plotted against species, with the speices sampled by many genes occur at the origin and they are ranked by the number of genes available. The latter file keeps the order of the trees and the taxa (alphabetical in both cases). This allows individual trees or taxa to be identified. A simple plotting script can be used to visualise these files.

===Fix Treefiles===
====Synopsis====
fix_treefiles.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing Tree files to fix
   --file            a file to fix
   --verbose         print verbose messages
</code>
====Options====
; '''--dir'''
:A directory which contains tree files that require fixing.
; '''--file'''
:A file to fix.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK

====Description====
'''Fix Treefiles''' is a simple script to fix some common problems with tree files created using various programs.

TreeView (Page, 1996): TreeView create a tree with the following description:

<code>
 UTREE * tree_1 = ((1,(2,(3,(4,5)))),(6,7));
</code>
UTREE * is not a supported part of the NEXUS format (as far as Bio::NEXUS (Hladish et. al., 2007) is concerned). stk_check_data.pl will report a formatting error with a file like this, so this script will fix this by replacingthe above with:

<code>
 tree_1 = [&u] ((1,(2,(3,(4,5)))),(6,7));
</code>
TNT (Goloboff et al, 2008): Outputed NEXUS tree files from TNT contain "begin trees ;" with a space bewteen the "trees" and the ";". Bio::NEXUS does not recognise this as the begining of a tree block and skips all the trees. stk_check_data.pl will pick this error up, removing the space.

The script works on either a file or directory.

====References====
Goloboff, P. A. and Farris, J. S. and Nixon, K. C., 2008. TNT, a free program for phylogenetic analysis. Cladistics 24 (5), 774-786. DOI: 10.1111/j.1096-0031.2008.00217.x

Hladish, T., Gopalan, V., Liang, C., Qiu, W., Yang, P., Stoltzfus, A., 2007. Bio::Nexus: A Perl API for the Nexus format for comparative biological data. BMC Bioinformatics 8, 191.

Page, R.D.M., 1996. Treeview: An application to display phylogenetic trees on personal computers. Computer Applications in the Biosciences 12, 357-358.

===Replace Genera===
====Synopsis====
stk_replace_genera.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing XML/Tree files. Required
   --verbose         print verbose messages. Sets C<--higher>, but substitutions are done.
   --higher          print higher taxa to screen. No substitutions will be done.
   --output          writes substitutions carried out to taxa substituion file (specify). No substitutions will be done.
</code>
====Options====
; '''--dir'''
:a directory which contains tree and/or XML files. '''Required'''.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--higher'''
:Prints the higher (non-specific) taxa to screen. No substitutions will be done
; '''--output'''
:Writes what substitutions will be done by the script to taxa substitution file. The default location is the data_directory/generic_subs.txt. This file is in the same file format used in [[#STK Replace Taxa|"STK Replace Taxa"]]. No substitutions will be done.

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK.

====Description====
'''Replace Genera''' is used when creating trees at specific level following the protocol of Davis (2008). Any taxon that are at generic level only are replaced by a polytomy of specific names from the rest of the data set.

An example is if a data set contains ''Gallus gallus'' and ''Gallus lafayetii'' will have ''Gallus'' replaced where it occurs as a generic taxon with a polytomy containing ''Gallus gallus'' and ''Gallus lafayetii''.

There is an assumption here that the XML and Treefiles live in the same directory. Therefore, specify the topmost directory where both can be reached. This won't work where, for example, there are two directories containing treefiles and there's one directory containing XML files underneath them, e.g. data_dir/ tree_original/ tree_processed/ xml_files/

Supplying <tt>data_dir</tt> will not work correctly.

Also, make sure that the data is consistant, i.e. run the Data Check script first and make sure the data passses.

===Replace Taxa===
====Synopsis====
stk_replace_taxa.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory in which replacements should occur 
   --file            file in which replacments should occur
   --taxa            file specifying which substitions to carry out            
   --old             taxon to be replaced
   --new             the new taxon. May be omitted, in which case old taxon is removed
   --verbose         verbose messages
   --reverse         the taxa file will contain taxa to keep with this flag on
</code>
One of <tt>dir</tt> or <tt>file</tt> is required. One of <tt>old</tt> or <tt>taxa</tt> is required.

====Options====
; '''--dir'''
:A directory which contains tree files that require replacing.
; '''--file'''
:The file (tree or XML) which has taxa that require replacing
; '''--taxa'''
:A file containing the substitutions to be made. See full documentation for details
; '''--old'''
:The old taxa to be replaced. Note that spaces in the taxon name must be replaced with _
; '''--new'''
:The new taxa. Note that spaces in the taxon name must be replaced with _. If blank, <tt>old</tt> will be deleted.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages

One of <tt>file</tt> or <tt>dir</tt> must be specified. Either <tt>taxa</tt> or <tt>old</tt> must also be specified. <tt>new</tt> can be ommitted to replace the original taxon.

====Description====
'''Replace Taxa''' replaces a single taxon or multiple taxa with either another taxon or multiple taxa as a polytomy. Replacement is done on both tree files and XML files (if present). If a file is specified only that file has the taxa replaced.

A taxon can be deleted by specifying a blank new taxon.

All taxon names are '''case sensitive'''.

Multiple replacements are achieved by supplying a "taxa file". This lists "old taxa" and "new taxa" on individual lines in the file, seperated by an "=" sign. For example: old_taxon = new_taxon old_taxon2 = new_taxon1, new_taxon2, new_taxon3 old_taxon3 =

The first line above does a simple substitution for <tt>old_taxon</tt> with <tt>new_taxon</tt>. The second line replaces <tt>old_taxon2</tt> with a polytomy of <tt>new_taxon1</tt>, <tt>new_taxon2</tt> and, <tt>new_taxon3</tt> in a tree file and a simple substitution in an XML file. The third line removes <tt>old_taxon3</tt>. Note there are spaces '''either''' side of the equals sign. These spaces '''must''' occur or the file is invalid.

In addition, the underscores (which are required for the tree file) are required here also. However, they will be replaced by spaces in the XML file.

The script works on either a single file or directory of files (including subdirectories).

'''Warning''': This script simply overwrites existing data - please back up data before using it.

====Examples====
Remove all instances of "MRPOutgroup" from a dataset

<code>
 perl stk_replace_taxa.pl --dir /home/data/phylogenies --old MRPOutgroup
</code>
Replace taxa_1 with taxa_2 within a single tree file

<code>
 perl stk_replace_taxa.pl --file /home/data/tree1.tre --old taxa_1 --new taxa_2
</code>
===Tree Permutation===
====Synopsis====
tree_permutation.pl [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --file            a file that contains polyphyletic taxa. REQUIRED
   --verbose         print verbose messages
   
</code>
====Options====
; '''--file'''
:a file which contains specially notation for polyphyly
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages

====Description====
'''Tree Permutation''' generates multilpe trees based on a single input tree. The input tree contains specially noted taxa which are polyphyletic or are, perhaps sub-species, which need to be removed.

Taxa are denoted by % symbol at the end of the name.

A single file is output.

===Search Data===
====Synopsis====
search_data [options]

<code>
 Options:
   --help            brief help message
   --man             full documentation
   --dir             directory containing XML or tree files. Required
   --verbose         print verbose messages
   --output          direct output to file instead of STDOUT
   --copy            copy all files resulting from search into a directory
   --charterm        a character term to search for 
   --anterm          an analysis term to search for
   --taxterm         a taxa to search for
   --anyterm         search for data for a term in either character, analysis, or taxa
   --fossil          none|all|some
   --only            restrict results to B<only> containing the search term. Character term only.
   --year            search for data from a particular year
</code>
====Options====
; '''--dir'''
:a directory which contains XML and/or tree files. '''Required'''.
; '''--help'''
:Print a brief help message and exits.
; '''--man'''
:Prints the manual page and exits.
; '''--verbose'''
:Print verbose messages
; '''--output'''
:Directs output to file instead of STDOUT. If a filename is given, output is directed there, otherwise it is directed to results.txt in the <tt>dir</tt>ectory.
; '''--copy'''
:Copy the files that are matched in the search to a new directory. If a directory name is given, the files are copied there. If this directory does not exist it is created. If no name is given, the files will be copied into a directory called <tt>seach_results</tt> in <tt>dir</tt>ectory.
; '''--charterm'''
:A character term to search for. Specify multiple <tt>--charterm</tt>s to perform OR searches
; '''--anterm'''
:An analysis term to search for. Specify multiple <tt>--anterm</tt>s to perform OR searches
; '''--taxterm'''
:A taxa to search for. Specify multiple <tt>--taxterm</tt>s to perform OR searches
; '''--anyterm'''
:A term to search for. Specify multiple <tt>--anyterm</tt>s to perform OR searches. This will search all of taxa, characters and analyses.
; '''--fossil'''
:Search for either <tt>none</tt>, <tt>some</tt> or <tt>all</tt> fossil-based data.
; '''--year'''
:Search for a particular year. Must be a four digit year.

If <tt>dir</tt> is specified all subdirectories will also be scanned, apart from those excluded by STK.

====Description====
'''Search Data''' allows the data directory to be searched for taxa, characters or analyses. The results can be printed to screen or the Tree/XML files copied into a new directory. The executable is useful for extracting subsets of data, for example all source trees generated using molecular data. XML files are required as the searches are carried on

Characters can be specified, e.g. cytb or rag1, or can be generic characters, e.g. Molecular or Morphological.

Taxonomic searches search only taxa contained in the data - this script does not understand taxonomy, so search for "Galliformes" will not return any results unless a tree contains a terminal node "Galliformes"

All searches are case insensitive. If multilpe terms are searchd an "OR" search will be carried out

==Bio::STK Module==
===STK - Supertree Tool Kit===
====Description====
'''STK''' is a collection of tools to prepare data ready for supertree contruction with PAUP* or similar. It relies on a file-based database based on tree and XML files. Functions include finding all tree files, xml files and summarising data. STK also comes with a number of scripts ready for use.

====Requires====
Perl 5.004, XML::Simple, Bio::NEXUS, File::Find, Carp, File::Copy;

====Feedback====
All feedback (bugs, feature enhancements, etc.) are all greatly appreciated.

====Authors====
Jon Hill (jon.hill@imperial.ac.uk) and Katie Davis (katie.davis@oum.ox.ac.uk)

====Methods====
=====find_tree_files=====
<code>
 Title   : find_tree_files
 Usage   : my @files = Bio::STK::find_tree_files($dir);
 Function: Finds all tree files in specified directory and all sub-directories
 Returns : Array of complete filenames
 Args    : $dir, else croaks.
 
</code>
=====find_xml_files=====
<code>
 Title   : find_xml_files
 Usage   : my @files = Bio::STK::find_xml_files($dir);
 Function: Finds all XML files in specified directory and all sub-directories
 Returns : Array of complete filenames
 Args    : $dir, else croaks.
</code>
=====read_tree_file=====
<code>
 Title   : read_tree_file
 Usage   : my $tree = read_tree_file($file);
 Function: Reads in a correctly formatted NEXUS tree file
 Returns : An array of Newick tree formatted strings;
 Args    : $file, else croaks.
</code>
=====read_xml_file=====
<code>
 Title   : read_xml_file
 Usage   : my $xml = read_xml_file($file);
 Function: Reads in an XML file
 Returns : The data structure from XML::Simple
 Args    : $file, else croaks. 
</code>
A sample data structure is:

<code>
 'Notes' => [
           '2,430bp of the mitochondrial genes ND2, ND3 and cyt b.'
         ],
 'Analysis' => [
              {
                'Type' => [
                            'MP'
                          ]
              }
            ],
 'Characters' => [
                {
                  'Other' => [
                               {}
                             ],
                  'Behavioural' => [
                                     {}
                                   ],
                  'Morphological' => [
                                       {}
                                     ],
                  'Molecular' => [
                                   {
                                     'Type' => [
                                                 'cytb',
                                                 'ND2',
                                                 'ND3'
                                               ],
                                     'number' => '3'
                                   }
                                 ]
                }
              ],
 'Taxa' => [
          {
            'List' => [
                        'Campylorhamphus falcularius',
                        'Campylorhamphus procurvoides',
                        'Campylorhamphus trochilirostris',
                        'Dendrexetastes rufigula',
                        'Dendrocolaptes certhia',
                        'Glyphorynchus spirurus',
                        'Hylexetastes perrotii',
                        'Lepidocolaptes albolineatus',
                        'Lepidocolaptes angustirostris',
                        'Lepidocolaptes fuscus',
                        'Nasica longirostris',
                        'Sittasomus griseicapillus',
                        'Xiphocolaptes promeropirhynchus',
                        'Xiphorhynchus erythropygius',
                        'Xiphorhynchus flavigaster',
                        'Xiphorhynchus guttatus',
                        'Xiphorhynchus kienerii',
                        'Xiphorhynchus lachrymosus',
                        'Xiphorhynchus obsoletus',
                        'Xiphorhynchus ocellatus',
                        'Xiphorhynchus pardalotus',
                        'Xiphorhynchus picus',
                        'Xiphorhynchus spixii',
                        'Xiphorhynchus susurrans',
                        'Xiphorhynchus triangularis'
                      ],
            'number' => '25',
            'fossil' => 'none'
          }
        ],
 'Source' => [
            {
              'Year' => [
                          '2000'
                        ],
              'Publisher' => [
                               {}
                             ],
              'Volume' => [
                            '119'
                          ],
              'Editor' => [
                            {}
                          ],
              'Pages' => [
                           '621-640'
                         ],
              'Booktitle' => [
                               {}
                             ],
              'Journal' => [
                             'Auk'
                           ],
              'Title' => [
                           'Molecular systematics and the role of the "Varzea"-"Terra-firme" ecotone in
 the diversification of Xiphorhynchus woodcreepers (Aves: Dendrocolaptidae).'
                         ],
              'Author' => [
                            'Bar, A.'
                          ]
            }
          ],
 'TreeFile' => [
              'tree1.tre'
            ]
</code>
You can access by using structures like this: <tt>@{$xml_contents-</tt>{Taxa}->[0]->{List}}>.

=====save_tree_file=====
<code>
 Title   : save_tree_file
 Usage   : save_tree_file($treefile, \@tree_string, \@tree_names);
 Function: Save an array of tree strings to file. Will overwrite any existing file without checking
 Returns : 1 if sucessful, 0 if not. CHECK THIS
 Args    : $treefile, @tree_string, @tree_names
</code>
This function does not check if the treefile already exists - this is the responsibility of the calling program.

=====save_tree_newick=====
<code>
 Title   : save_tree_newick
 Usage   : save_tree_newick($treefile, \@tree_string, \@tree_names)
 Function: Save an array of tree strings to file. Will overwrite any existing file without checking
 Returns : 1 if sucessful, 0 if not. CHECK THIS
 Args    : $treefile, \@tree_string, \@tree_names
</code>
This function does not check if the treefile already exists - this is the responsibility of the calling program. The purpose of this function is to save a tree file in "raw" Newick format, i.e. without any other BLOCKS apart from TREE. This is the simplist form of the treefile and hence the most compatible with other software.

=====save_xml_file=====
<code>
 Title   : save_xml_file
 Usage   : save_xml_file($xmlfile, $xml_hash);
 Function: Save a XML hash to file. Will overwrite any existing file without checking
 Returns : 1 if sucessful, 0 if not. CHECK THIS
 Args    : $xmlfile, $xml_hash
</code>
This function does not check if the xml file already exists - this is the responsibility of the calling program.

$xml_hash is the same object as returned by [[read_xml_file]].

=====read_taxa_file=====
<code>
 Title   : read_taxa_file
 Usage   : $nSubs = read_taxa_file($taxafile, \@old_taxa, \@new_taxa);
 Function: Read in a taxa substitution file, populating @old_taxa and @new_taxa.
 Returns : The number of substitutions.
 Args    : $taxafile, \@old_taxa, \@new_taxa
</code>
old_taxa contains each old taxa, with a corresponding entry in new_taxa, i.e.

<code>
 old_taxa[0] => new_taxa[0]
 old_taxa[1] => new_taxa[1]
 etc
</code>
new_taxa will contain a string which is a comma-seperated list of all taxa that should replace the old_taxa

=====taxa_from_tree=====
<code>
 Title   : taxa_from_tree
 Usage   : my @taxa = taxa_from_tree($tree);
 Function: Obtains unique taxa from a tree file or tree string
 Returns : An array of taxa names or undef if error
 Args    : $tree, else croaks.
</code>
$tree can be a tree file or a NEXUS formatted tree string

=====get_taxa_list=====
<code>
 Title   : get_taxa_list
 Usage   : get_taxa_list($dir);
 Function: Gets a list of unique taxa from a directory (and subdirectories) of tree files.
 Returns : An array of unique names
 Args    : $dir, else croaks.
</code>
=====taxa_from_xml=====
<code>
 Title   : taxa_from_xml
 Usage   : my @taxa = taxa_from_xml($input)
 Function: Gets the taxa containedin XML file or XML hash
 Returns : An array of strings containing the taxa
 Args    : $input, else croaks. 
</code>
Note that input can be a filename or XML hash.

=====get_analysis=====
<code>
 Title   : get_analysis
 Usage   : $analysis = get_analysis($xmlfile);
 Function: Gets the type of analysis used from an XML file
 Returns : A string containing the type of analysis used
 Args    : $xmlfile, else croaks.
</code>
=====get_year=====
<code>
 Title   : get_year
 Usage   : $year = get_year($xmlfile);
 Function: Gets the year from the bibliographic data
 Returns : An integer containing the year the data were published
 Args    : $xmlfile, else croaks.
</code>
=====get_source_data=====
<code>
 Title   : get_source_data
 Usage   : %source = get_source_data($xmlfile);
 Function: Gets the tsource information from an XML file
 Returns : A hash containing the source information (journal, title, etc)
 Args    : $xmlfile, else croaks. 
</code>
The output from get_source_data returns a Hash with the following structure:

<code>
 my %expected = (year => '2000',
                volume => '119',
                pages => '621-640',
                journal => 'Auk',
                title => 'Molecular systematics and the role of the "Varzea"-"Terra-firme" ecotone in the diversification of Xiphorhynchus woodcreepers (Aves: Dendrocolaptidae).',
                author => 'Bar, A.',
                booktitle => 'Some title',
                editor => 'Some editors',
                publisher => 'Some publisher',
                );
</code>
Access as a normal Hash.

=====get_treefile=====
<code>
 Title   : get_treefile
 Usage   : get_treefile($xmlfile);
 Function: Get the tree files that this XML file is for
 Returns : A string containing the B<relative> path of the tree file
 Args    : $xmlfile, else croaks. 
</code>
=====get_characters=====
<code>
 Title   : get_characters
 Usage   : get_characters($xmlfile);
 Function: Get the chracters used in this analysis
 Returns : A Hash containing the character types
 Args    : $xmlfile, else croaks.
</code>
The output from get_characters returns a Hash with the following structure:

<code>
 my %expected = (Molecular => [ "cytb", "ND2", "ND3" ],
                Morphological => [ "feathers" ],
                Other => [ "stuff" ],
                Behavioural => [ "stuff", "more stuff" ],
                );
</code>
Access as a normal Hash and then array notation, i.e. <tt>$expected{molecular}[2]</tt>;

=====check_tree_file=====
<code>
 Title   : check_tree_file
 Usage   : check_tree_file($file);
 Function: Test that a tree file is a valid NEXUS file
 Returns : 1 on success, 0 on failure - REMEMBER TO CHECK THIS!
 Args    : $file, else croaks.
</code>
=====replace_taxon_tree=====
<code>
 Title   : replace_taxon_tree
 Usage   : replace_taxon_tree(\@treeString, $old_taxon, \@new_taxa);
 Function: Swap a taxon in a tree with a new one within an array of tree strings. Leave new_taxon blank to remove old_taxon
 Returns : 1 on success, 0 on failure. \@treeString contains new array of trees
 Args    : \@treeString, $old_taxon, \@new_taxa. new_taxa may be null, in which case old_taxon is removed.
           @new_taxon is an array of names - if this is longer than 1 element, then the old_taxon is
           replaced with a polytomy
 
</code>
Note that this function uses "pass by reference" - remember to add the slashes

=====delete_taxa=====
<code>
 Title   : delete_taxa
 Usage   : delete_taxa(\@treeString, \@to_delete);
 Function: Remove multilple taxa from a tree
 Returns : 1 on success, 0 on failure. \@treeString contains new array of trees
 Args    : \@treeString, \@to_delete
           \@treeString is a reference to an array of tree strings
           \@to_delete is an array of names
 
</code>
Note that this function uses "pass by reference" - remember to add the slashes

=====replace_taxon_xml=====
<code>
 Title   : replace_taxon_xml
 Usage   : replace_taxon_xml($xml_data, $old_taxon, @new_taxon);
 Function: Swap a taxon in a XML data structure with a new one. Leave new_taxon blank to remove old_taxon
 Returns : 1 on success, 0 on failure. $xml_data is altered.
 Args    : $xml_data, $old_taxon, @new_taxon. new_taxon may be null, in which case old_taxon is removed.
           @new_taxon is an array of names - if this is longer than 1 element, then the old_taxon is
           replaced with a polytomy
</code>
=====tree_equals=====
<code>
 Title   : tree_equals
 Usage   : tree_equals($tree1, $tree2);
 Function: Compare tree1 to tree2 and test for equality
 Returns : 1 if equal, 0 if not. CHECK THIS
 Args    : $tree1, $tree2 - both NEWICK formatted strings.
</code>
=====tree_contains=====
<code>
 Title   : tree_contains
 Usage   : tree_contains($taxon,$tree_data);
 Function: Check if $tree_data contains $taxon. Case insensitive. Accounts for _ or spaces in name
 Returns : 1 if it does, 0 if not. CHECK THIS
 Args    : $taxon, $tree_data
</code>
Note that $tree_data can be a tree file or a tree string

=====xml_from_year=====
<code>
 Title   : xml_from_year
 Usage   : xml_from_year($year,$xml_file);
 Function: Check if $xml_file is from $year.
 Returns : 1 if it does, 0 if not. CHECK THIS
 Args    : $year, $xml_file
</code>
$xml_file can be a file or XML data structure

=====xml_contains_taxon=====
<code>
 Title   : xml_contains_taxon
 Usage   : xml_contains_taxon($taxon,$xml_file);
 Function: Check if $xml_file contains $taxon. Case insensitive. Accounts for _ or spaces in name
 Returns : 1 if it does, 0 if not. CHECK THIS
 Args    : $taxon, $xml_file
</code>
$xml_file can be a file or XML data structure

=====xml_contains_analysis=====
<code>
 Title   : xml_contains_analysis
 Usage   : xml_contains_analysis($analysis,$xml);
 Function: Check if $xml contains $analysis
 Returns : 1 if it does, 0 if not. CHECK THIS
 Args    : $analysis, $xml
</code>
$xml can be a file or XML data structure

=====xml_contains_character=====
<code>
 Title   : xml_contains_character
 Usage   : xml_contains_character($character,$xml,$only);
 Function: Check if $xml contains $character
 Returns : 1 if it does, 0 if not. CHECK THIS
 Args    : $character, $xml
</code>
$xml can be a file or XML data structure. <tt>$character</tt> can be a specific charater type, e.g. 'cytb' or can be one of "Molecular", "Morphological", "Behavioural" or "Other".

=====contains_data=====
<code>
 Title   : contains_data
 Usage   : contains_data(\@data,$xml)
 Function: Check if $xml contains $data. $data is case insensitive, but spaces in taxa names are not accounted for
 Returns : 1 if it does, 0 if not. CHECK THIS
 Args    : $data, $xml
 
</code>
<tt>$data</tt> is a string or array of strings containing the data to be searched for within the <tt>$xml</tt>. <tt>$data</tt> can be one of:

; Characters =item Analysis =item Taxa
If <tt>$data</tt> is an array an OR search is carried out.

<tt>$xml</tt> can be either a XML has (from read_xml_file) or a file name.

=====contains_fossils=====
<code>
 Title   : contains_fossils
 Usage   : contains_fossils($xmlfile);
 Function: Tests if XML file contains fossil taxa
 Returns : 0 - no fossils, 1 all fossils, 2 mixed, -1 incorrect info
 Args    : $xmlfile, else croaks. 
</code>
=====get_short_study_name=====
<code>
 Title   : get_short_study_name
 Usage   : get_short_study_name(@xmlFileList);
 Function: Return a unique (for the dataset defined by the input array) name for a source tree study
 Returns : A linked list containing files (key) and the short name (value)
 Args    : @xmlfile
</code>
Call this function before the main loop over XML files to generate the short name. You can then use the resulting hash to grab the short name for whichever file you're currently dealing with.

